flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/TestingPhysicalSlotProvider.java

```diff
@@ -0,0 +1,193 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.scheduler;
+
+import org.apache.flink.runtime.clusterframework.types.AllocationID;
+import org.apache.flink.runtime.clusterframework.types.ResourceProfile;
+import org.apache.flink.runtime.concurrent.FutureUtils;
+import org.apache.flink.runtime.executiongraph.utils.SimpleAckingTaskManagerGateway;
+import org.apache.flink.runtime.jobmanager.scheduler.NoResourceAvailableException;
+import org.apache.flink.runtime.jobmanager.slots.TaskManagerGateway;
+import org.apache.flink.runtime.jobmaster.SlotRequestId;
+import org.apache.flink.runtime.jobmaster.slotpool.PhysicalSlotProvider;
+import org.apache.flink.runtime.jobmaster.slotpool.PhysicalSlotRequest;
+import org.apache.flink.runtime.taskmanager.LocalTaskManagerLocation;
+import org.apache.flink.runtime.taskmanager.TaskManagerLocation;
+import org.apache.flink.util.Preconditions;
+import org.apache.flink.util.function.TriFunctionWithException;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A {@link PhysicalSlotProvider} implementation that can be used in tests.
+ */
+public class TestingPhysicalSlotProvider implements PhysicalSlotProvider {
+
+	private final Map<SlotRequestId, PhysicalSlotRequest> requests;
+	private final Map<SlotRequestId, CompletableFuture<TestingPhysicalSlot>> responses;
+	private final Map<SlotRequestId, Throwable> cancellations;
+
+	private final TaskManagerLocation taskManagerLocation;
+	private final TaskManagerGateway taskManagerGateway;
+
+	private final TriFunctionWithException<TaskManagerLocation, TaskManagerGateway, ResourceProfile, CompletableFuture<TestingPhysicalSlot>, Throwable> physicalSlotCreator;
+
+	private TestingPhysicalSlotProvider(TriFunctionWithException<TaskManagerLocation, TaskManagerGateway, ResourceProfile, CompletableFuture<TestingPhysicalSlot>, Throwable> physicalSlotCreator, TaskManagerLocation taskManagerLocation, TaskManagerGateway taskManagerGateway) {
+		this.physicalSlotCreator = physicalSlotCreator;
+
+		this.taskManagerLocation = taskManagerLocation;
+		this.taskManagerGateway = taskManagerGateway;
+
+		this.requests = new HashMap<>();
+		this.responses = new HashMap<>();
+		this.cancellations = new HashMap<>();
+	}
+
+	@Override
+	public CompletableFuture<PhysicalSlotRequest.Result> allocatePhysicalSlot(PhysicalSlotRequest physicalSlotRequest) {
+		SlotRequestId slotRequestId = physicalSlotRequest.getSlotRequestId();
+		requests.put(slotRequestId, physicalSlotRequest);
+		CompletableFuture<TestingPhysicalSlot> resultFuture;
+		try {
+			resultFuture = physicalSlotCreator.apply(
+				taskManagerLocation,
+				taskManagerGateway,
+				physicalSlotRequest.getSlotProfile().getPhysicalSlotResourceProfile());
```

Comment:
I am still a bit sceptical that `taskManagerLocation` and `taskManagerGateway` belong to the `TestingPhysicalSlotProvider`. I think they rather belong to the `physicalSlotCreator`. An indicator is that we now have 2 ways of how to specify the `TaskManagerLocation`. 1) set a custom `physicalSlotCreator` which is given the location 2) set the location for the `TestingPhysicalSlotProvider`. I think this is an indicator that there is a bit of ambiguity in the separation of concerns.

Reply:
Now I get it. Thanks for clarification. I also removed the Builder as part of this refactoring since we only have the creator function left as a parameter.